### ISSUE TITLE
RawKeyEvent and RawKeyboard, et al should be deprecated and removed

### ISSUE THREAD

**Issue by gspencergoog** [flutter/flutter#136419]
## Description

Several years ago, we embarked on a project to improve the keyboard key handling for Flutter.  The new system has been in production use for a while now (at least a couple of years).

However, the legacy interfaces are still present, and on the framework side, this makes for a confusing story for our users, and complicates the code. On the engine side, we are still producing and handling both kinds of keyboard events, which complicates the code and is less efficient.

This document describes the steps that need to be taken in order to complete the deprecation and removal of the old “Raw” APIs and code.

## Affected APIs

The APIs that need to be deprecated and eventually removed are:

- RawKeyboard
- RawKeyEvent
- RawKeyDownEvent
- RawKeyUpEvent
- RawKeyEventData
- RawKeyEventHandler
- RawKeyEventDataAndroid
- RawKeyEventDataLinux
- RawKeyEventDataIos
- RawKeyEventDataMacOs
- RawKeyEventDataWindows
- RawKeyEventDataFuchsia
- RawKeyEventDataWeb
- RawKeyEventData
- FocusOnKeyCallback
- RawKeyboardListener
- KeyEventHandler.handleRawKeyMessage
- KeyboardSide
- ModifierKey
- KeyDataTransitMode
- KeySimulatorTransitModeVariant
- Focus.onKey
- FocusScope.onKey
- FocusNode.onKey
- FocusScopeNode.onKey
- debugKeyEventSimulatorTransitModeOverride

Things that still need to be migrated to use only KeyEvent instead of RawKeyEvent (in some cases these currently work with both, and will eventually only need to work with KeyEvent):

- ShortcutActivator.accepts
- ShortcutActivator.isActivatedBy
- simulateKeyDownEvent
- simulateKeyUpEvent
- KeyEventSimulator
- _Many tests_

Functionality that existed in RawKeyEvent that no longer exists after switch to KeyEvent:

- Convenience functions for determining if a key is pressed.
	- bool isKeyPressed(LogicalKeyboardKey key)
	- bool get isControlPressed
	- bool get isShiftPressed
	- bool get isAltPressed
	- bool get isMetaPressed

Things that need to be removed:

- All related private APIs
- Key code generation that refers to RawKeyEvent
- Engine side code that generates RawKeyEvent events.
- Documentation references to all deprecated APIs, both in API docs and website docs.

## Deprecation Timeline

Q4 2023: Deprecate all the API listed above as deprecated.
Q4 2024: Remove RawKeyboard and RawKeyEvent, et al. from the Flutter Framework.
Q4 2024: Remove related code from the Flutter Engine.

**Comment by gspencergoog** [comment]
For the migration away from `RawKeyEvent`, we need to make `ShortcutActivator.accepts` handle `KeyEvent` events. The options appear to be:

	1) add a new `acceptsKeyEvent` method that takes a `KeyEvent`, and make people override it (it will break all subclasses when it is added), or 
	2) create a sealed base class that only `RawKeyEvent` and `KeyEvent` derive from (`GenericKeyEvent` perhaps) and make the `ShortcutActivator.accepts` method take a `covariant GenericKeyEvent`, and then provide two implementations conditional on which type was given. Later, both `RawKeyEvent` and `GenericKeyEvent` would be removed.

The up side of the first method is that I don't need to relate the two key event classes, but the down side is a huge breaking change up front when the new method is added.

The up side of the second way is that we don't need to rename the method, and we don't break people's subclasses, but the down side is that we'll break all the subclasses anyhow when we remove `GenericKeyEvent`, or at runtime when we switch to sending only `KeyEvent`s.

We could just leave `GenericKeyEvent` in place after removing the deprecated `RawKeyEvent`, but it would have no members, and only one subclass (`KeyEvent`), so it's really pretty useless. We could start its deprecation period when we remove `RawKeyEvent`, I guess, and then remove it two years from now...

The other part of the problem is when to switch over from sending `RawKeyEvent` instances to `ShortcutActivator.accepts` to sending `KeyEvent` instances.  If I go with the sealed base class path, then it will just break a lot of code at runtime, and static failures are preferable to runtime failures, of course.

Here's what I might try:

1) Introduce the sealed base class (`GenericKeyEvent`) and update `ShortcutActivator.accepts`:
	- Introduce the sealed base class `GenericKeyEvent` (which will mean moving implementation of RawKeyEvent in the framework into the same file as `KeyEvent`, but it'll be deprecated and removed eventually).
	- Update the `ShortcutActivator.accepts` method to accept a `covariant GenericKeyEvent.`
	- Maintain the existing implementation of `ShortcutActivator.accepts` for `RawKeyEvent`.
	- Add a new implementation for `ShortcutActivator.accepts` that handles `KeyEvent` objects.
	- Provide a global setting that changes wha

2) Deprecate `RawKeyEvent`:
	- Mark the `RawKeyEvent` class as deprecated.
	- Provide a migration guide for developers to switch from `RawKeyEvent` to `KeyEvent`.
 
3) Transition to `KeyEvent`:
	- Update the code that sends key events to `ShortcutActivator.accepts` to use KeyEvent instead of `RawKeyEvent`.
	- Maintain the existing implementation for `RawKeyEvent` in `ShortcutActivator.accepts` for a deprecation period.

4) Remove `RawKeyEvent`:
	- After the deprecation period, remove the `RawKeyEvent` class.
	- Remove the `RawKeyEvent`-specific implementation from `ShortcutActivator.accepts`.

**Comment by gspencergoog** [comment]
After talking with Michael, it seems like method 1, above might be the best way to go.  Yes, it causes a breaking change for people who subclass `ShortcutActivator` directly, but most people don't do that.

I'm going to give that a try.

**Comment by rydmike** [comment]
This change will require a lot of rework in several repos. Been using things like `RawKeyboardListener` and e.g. `RawKeyEventDataMacOs`, `RawKeyEventDataIos`, since that was pretty much the only way to get what was needed done, when it was written **many** years ago.

I've seen there are some newer ways, but no real idea what they are. Yes these old APIs are fairly low level, but once setup they worked worked very well, so there was no need to fix them or look into the newer APIs, since they were not broken.

I hope good and very in-depth guides on how to replace all the deprecated APIs above will be provided. At least I have no clue where to even begin look for replacements.

**Comment by bernaferrari** [comment]
I'm all in for this change, except for the fact that shortcut is super painful. I hate it. Isn't there a better way? I understand its benefits, but the whole intents mixed with commands that are basically static (actions, bindings) make everything so hard. Could you use shortcuts to, say, control a character with WASD, while having everything in the same function? I miss a dynamic approach. Could you swap the behavior depending on user locale? I don't think so. Things are way too limited for my taste, which is the reason I usually don't use shortcuts in my apps.

**Comment by gspencergoog** [comment]
@rydmike Yes, sorry for the churn, but the raw methods have been marked in the documentation as "going to be deprecated" for a while now (years), and we really need to remove the duplicate system to reduce the tech debt of maintaining them both, or it will just grow over time (as it already has).

There will definitely be an extensive migration guide and, for the things we use it for, `dart fix` solutions.

The replacements are generally found linked in the existing class documentation.  For the ones you mentioned, `RawKeyboardListener` is replaced by `KeyboardListener`, and (hopefully) the information in the platform specific classes you mentioned is available in `KeyEvent`, because the platform-specific data classes are being phased out. If not, please let us know what is missing.

**Comment by gspencergoog** [comment]
Okay, with the commit of #136677, the deprecations are in, starting the one year countdown to their removal.

As mentioned in the issue description above, the migration guide for this is here: https://docs.flutter.dev/release/breaking-changes/key-event-migration

**Comment by aymswick** [comment]
We recently migrated our keyboard shortcut listener's `FocusNode` to use `onKeyEvent` instead of `onKey`. Currently, using `tester.sendKeyUpEvent` does not appear to be triggering our handler. Using `tester.sendKeyDownEvent` does work as expected. Is this to be expected until the final checklist items re: simulated events is completed? Thanks in advance, just want to document test failures if we anticipate them passing as-is when this issue is completed.

**Comment by gspencergoog** [comment]
@aymswick Could you please file an issue for what you described? (Feel free to cc me.)

**Comment by biancashouse** [comment]
Flutter 3.19.1
OSX Venture

Flutter Web

When I refactor from using RawKeyboardListener to KeyboardListener, no key events get triggered.
I'm using the listener to implement a way to close Overlays with the Escape key.

**Comment by spydon** [comment]
> .. the information in the platform specific classes you mentioned is available in KeyEvent, because the platform-specific data classes are being phased out. If not, please let us know what is missing.

@gspencergoog In the Ubuntu installer we were are using the following to be able to let the user detect which keyboard layout their keyboard has:
```
void _handleKey(RawKeyEvent event) {
  final data = event.data as RawKeyEventDataLinux;
  final scanCode = data.scanCode;
  ...
}
```

Is this not possible to do in the new API?

**Comment by spydon** [comment]
I solved it with manual mapping for now (if anyone else needs this).
```dart
  bool _handleKey(KeyEvent event) {
    if (event is! KeyDownEvent) return false;
    final hidCode = event.physicalKey.usbHidUsage & 0xFFFF;
    final scanCode = _usbHidToScanCode[hidCode] ?? -1;
    widget._onKeyPress?.call(scanCode);
    return true;
  }

...

final _usbHidToScanCode = {
  0x01: 0x00FF, // Overrun Error
  0x02: 0x00FC, // POST Fail
  0x04: 0x001E, // a A
  0x05: 0x0030, // b B
  0x06: 0x002E, // c C
  0x07: 0x0020, // d D
  0x08: 0x0012, // e E
  0x09: 0x0021, // f F
  0x0A: 0x0022, // g G
  0x0B: 0x0023, // h H
  0x0C: 0x0017, // i I
  0x0D: 0x0024, // j J
  0x0E: 0x0025, // k K
  0x0F: 0x0026, // l L
  0x10: 0x0032, // m M
  0x11: 0x0031, // n N
  0x12: 0x0018, // o O
  0x13: 0x0019, // p P
  0x14: 0x0010, // q Q
  0x15: 0x0013, // r R
  0x16: 0x001F, // s S
  0x17: 0x0014, // t T
  0x18: 0x0016, // u U
  0x19: 0x002F, // v V
  0x1A: 0x0011, // w W
  0x1B: 0x002D, // x X
  0x1C: 0x0015, // y Y
  0x1D: 0x002C, // z Z
  0x1E: 0x0002, // 1 !
  0x1F: 0x0003, // 2 @
  0x20: 0x0004, // 3 #
  0x21: 0x0005, // 4 $
  0x22: 0x0006, // 5 %
  0x23: 0x0007, // 6 ^
  0x24: 0x0008, // 7 &
  0x25: 0x0009, // 8 *
  0x26: 0x000A, // 9 (
  0x27: 0x000B, // 0 )
  0x28: 0x001C, // Return
  0x29: 0x0001, // Escape
  0x2A: 0x000E, // Backspace
  0x2B: 0x000F, // Tab
  0x2C: 0x0039, // Space
  0x2D: 0x000C, // - _
  0x2E: 0x000D, // = +
  0x2F: 0x001A, // [ {
  0x30: 0x001B, // ] }
  0x31: 0x002B, // \ |
  0x33: 0x0027, // ; :
  0x34: 0x0028, // ' "
  0x35: 0x0029, // ` ~
  0x36: 0x0033, // , <
  0x37: 0x0034, // . >
  0x38: 0x0035, // / ?
  0x39: 0x003A, // Caps Lock
  0x3A: 0x003B, // F1
  0x3B: 0x003C, // F2
  0x3C: 0x003D, // F3
  0x3D: 0x003E, // F4
  0x3E: 0x003F, // F5
  0x3F: 0x0040, // F6
  0x40: 0x0041, // F7
  0x41: 0x0042, // F8
  0x42: 0x0043, // F9
  0x43: 0x0044, // F10
  0x44: 0x0057, // F11
  0x45: 0x0058, // F12
  0x46: 0xE037, // Print Screen (Note 1)
  0x47: 0x0046, // Scroll Lock
  0x49: 0xE052, // Insert (Note 1)
  0x4A: 0xE047, // Home (Note 1)
  0x4B: 0xE049, // Page Up (Note 1)
  0x4C: 0xE053, // Delete (Note 1)
  0x4D: 0xE04F, // End (Note 1)
  0x4E: 0xE051, // Page Down (Note 1)
  0x4F: 0xE04D, // Right Arrow (Note 1)
  0x50: 0xE04B, // Left Arrow (Note 1)
  0x51: 0xE050, // Down Arrow (Note 1)
  0x52: 0xE048, // Up Arrow (Note 1)
  0x53: 0x0045, // Num Lock
  0x54: 0xE035, // Keypad / (Note 1)
  0x55: 0x0037, // Keypad *
  0x56: 0x004A, // Keypad -
  0x57: 0x004E, // Keypad +
  0x58: 0xE01C, // Keypad Enter
  0x59: 0x004F, // Keypad 1 End
  0x5A: 0x0050, // Keypad 2 Down
  0x5B: 0x0051, // Keypad 3 PageDn
  0x5C: 0x004B, // Keypad 4 Left
  0x5D: 0x004C, // Keypad 5
  0x5E: 0x004D, // Keypad 6 Right
  0x5F: 0x0047, // Keypad 7 Home
  0x60: 0x0048, // Keypad 8 Up
  0x61: 0x0049, // Keypad 9 PageUp
  0x62: 0x0052, // Keypad 0 Insert
  0x63: 0x0053, // Keypad . Delete
  0x64: 0x0056, // Europe 2 (Note 2)
  0x65: 0xE05D, // App
  0x67: 0x0059, // Keypad =
  0x68: 0x005D, // F13
  0x69: 0x005E, // F14
  0x6A: 0x005F, // F15
  0x85: 0x007E, // Keypad , (Brazilian Keypad .)
  0x87: 0x0073, // Keyboard Int'l 1 ろ (Ro)
  0x88: 0x0070, // Keyboard Int'l 2 かたかな ひらがな ローマ字 (Katakana/Hiragana)
  0x89: 0x007D, // Keyboard Int'l 3 ￥ (Yen)
  0x8A: 0x0079, // Keyboard Int'l 4 前候補 変換 (次候補) 全候補 (Henkan)
  0x8B: 0x007B, // Keyboard Int'l 5 無変換 (Muhenkan)
  0x8C: 0x005C, // Keyboard Int'l 6 (PC9800 Keypad , )
  0x90: 0x00F2, // Keyboard Lang 1 한/영 (Hanguel/English)
  0x91: 0x00F1, // Keyboard Lang 2 한자 (Hanja)
  0x92: 0x0078, // Keyboard Lang 3 かたかな (Katakana)
  0x93: 0x0077, // Keyboard Lang 4 ひらがな (Hiragana)
  0x94: 0x0076, // Keyboard Lang 5 半角/全角 (Zenkaku/Hankaku)
  0xE0: 0x001D, // Left Control
  0xE1: 0x002A, // Left Shift
  0xE2: 0x0038, // Left Alt
  0xE3: 0xE05B, // Left GUI
  0xE4: 0xE01D, // Right Control
  0xE5: 0x0036, // Right Shift
  0xE6: 0xE038, // Right Alt
  0xE7: 0xE05C, // Right GUI
};
```

**Comment by gspencergoog** [comment]
> Is this not possible to do in the new API?

Correct, this isn't possible in the new API, since it's more platform independent. If detecting the layout is something that makes sense in a cross platform way, we can work on figuring out a way to represent that.

**Comment by spydon** [comment]
> Correct, this isn't possible in the new API, since it's more platform independent. If detecting the layout is something that makes sense in a cross platform way, we can work on figuring out a way to represent that.

I can't come up with another use-case than ours (OS installer), or possibly if you have like a front-end for OS keyboard settings, so it's probably not worth spending time on unless others see the need for it. We'll be fine with the manual mapping. :)

**Comment by deskangel** [comment]
How can I migrate the following codes without rewriting the whole key binding logic under the new system?

```dart

  Keybinding({
    this.key = 'v',
    this.modifiers = RawKeyEventDataMacOs.modifierCommand | RawKeyEventDataMacOs.modifierLeftShift,
  });

  Map<String, dynamic> toJson() => {
        'key': this.key,
        'modifiers': this.modifiers,
      };

  Keybinding.fromJson(Map<String, dynamic> json) {
    key = json['key'] ?? 'v';
    modifiers = json['modifiers'] ?? RawKeyEventDataMacOs.modifierCommand | RawKeyEventDataMacOs.modifierLeftShift;
  }
  Keybinding.fromJsonString(String content) {
    var json = jsonDecode(content);
    key = json['key'] ?? 'v';
    modifiers = json['modifiers'] ?? RawKeyEventDataMacOs.modifierCommand | RawKeyEventDataMacOs.modifierLeftShift;
  }

  @override
  String toString() {
    var isCmdPressed =
        ((this.modifiers & RawKeyEventDataMacOs.modifierCommand) == RawKeyEventDataMacOs.modifierCommand);
    var isCtrlPressed =
        ((this.modifiers & RawKeyEventDataMacOs.modifierControl) == RawKeyEventDataMacOs.modifierControl);
    var isOptPressed = ((this.modifiers & RawKeyEventDataMacOs.modifierOption) == RawKeyEventDataMacOs.modifierOption);
    var isShiftPressed = ((this.modifiers & RawKeyEventDataMacOs.modifierShift) == RawKeyEventDataMacOs.modifierShift);

    String content = '';
    if (isCmdPressed) {
      content += 'CMD ';
    }
    if (isCtrlPressed) {
      content += 'CTRL ';
    }
    if (isOptPressed) {
      content += 'OPT ';
    }
    if (isShiftPressed) {
      content += 'SHIFT ';
    }
    if (this.key == ' ') {
      content += 'SPACE';
    } else {
      content += this.key.toUpperCase();
    }

    return content;
  }
```

**Comment by gspencergoog** [comment]
@deskangel Unfortunately, you will have to rewrite your logic, since the modifier mask is no longer provided.  There are, however, some helper functions that will help you do it: Look at [`HardwareKeyboard.isMetaPressed`](https://main-api.flutter.dev/flutter/services/HardwareKeyboard/isMetaPressed.html), etc.

**Comment by deskangel** [comment]
This is what I am concerned about. This isn't the first time I've had to
rewrite all the codes related to the keyboard handling without seeing any
benefits. Are the breaking changes caused by the removal absolutely
necessary, or is it just for the sake of tidiness or clean?

Best regards,
William Hsueh

Greg Spencer ***@***.***> 于 2024年3月26日周二 00:33写道：

**Comment by gspencergoog** [comment]
It is necessary in order to reduce the maintenance cost of maintaining the system, and to reduce the complexity of the system for stability and quality. It's not just because we want it "tidy", it's because we want it to be well engineered and stable, without being overly complex to understand.

**Comment by konrad321** [comment]
I just noticed that HardwareKeyboard.instance.isShiftPressed doesn't work on Android, whereas the RawKeyEvent event.isShiftPressed did work. This applies both to hardware and virtual keyboards. Is this the intended behaviour to drop support for isShiftPressed on Android or should a file a bug for this? I used it to distinguish Enter and Shift+Enter for some cases.

**Comment by gspencergoog** [comment]
@konrad321 No, that's not intended behavior, although it is intended that it doesn't work with virtual keyboards (because they don't generally generate events on Android). I don't think that ever worked for virtual keyboards.  Could you please file a separate issue for that with details?

**Comment by konrad321** [comment]
There is also a macOS issue with the new `HardwareKeyboard` which I reported in https://github.com/flutter/flutter/issues/148936 many months ago. I'm also stuck with the old `RawKeyboardListener` until it is resolved.
